### PR TITLE
http3: Add support for HTTP/3 METADATA

### PR DESCRIPTION
**Pull request by RyanTheOptimist**
http3: Add support for HTTP/3 METADATA

Adds a new `allow_metadata` option to Http3ProtocolOptions.

Risk Level: Low, protected by new config option
Testing: New integration tests
Docs Changes: N/A
Release Notes: Updated

### CODE REVIEW

File: api/envoy/config/core/v3/protocol.proto
Comment (abeyad) [inline api/envoy/config/core/v3/protocol.proto:641]: super nit: s/envoy/Envoy

File: api/envoy/config/core/v3/protocol.proto
Comment (abeyad) [inline api/envoy/config/core/v3/protocol.proto:644]: nit: perhaps "Allow HTTP/3 METADATA frames".  `metadata` is kind of vague and could mean many different things.

File: api/envoy/config/core/v3/protocol.proto
Comment (RyanTheOptimist) [reply api/envoy/config/core/v3/protocol.proto]: Done! (Copied this from the h2 section, so I changed it there as well).

File: api/envoy/config/core/v3/protocol.proto
Comment (RyanTheOptimist) [reply api/envoy/config/core/v3/protocol.proto]: Done! (ditto)

**Review by abeyad — DISMISSED** [review]
/lgtm api

File: api/envoy/config/core/v3/protocol.proto
Comment (abeyad) [reply api/envoy/config/core/v3/protocol.proto]: thanks Ryan!

**Comment by RyanTheOptimist** [comment]
Sorry! I keep forgetting that Envoy does automatic assignment. I think the code in this PR is basically ready for review, but it depends on some un-commited QUICHE changes that I manually passed in so it won't yet pass CI

**Comment by RyanTheOptimist** [comment]
/assign @danzh2010 

Ok, this is ready for review, I hope. There's a QUICHE update in this as well, but I'll land that separately once #32642 lands.

**Comment by RyanTheOptimist** [comment]
Now that #32650 has landed, this PR does not change deps, so I'll remove that label.

File: source/common/quic/envoy_quic_stream.cc
Comment (danzh2010) [inline source/common/quic/envoy_quic_stream.cc:160]: Why is this ASSERT held?

File: source/common/quic/envoy_quic_stream.cc
Comment (danzh2010) [inline source/common/quic/envoy_quic_stream.cc:126]: Variables at function scope shouldn't be suffixed with underscore.

File: source/common/quic/envoy_quic_stream.cc
Comment (danzh2010) [inline source/common/quic/envoy_quic_stream.cc:172]: Connection might be closed during write loop. Maybe early return in that case?

**Review by danzh2010 — COMMENTED** [review]
Thanks for making this change! Only a few nits left.

File: source/common/quic/envoy_quic_stream.cc
Comment (RyanTheOptimist) [reply source/common/quic/envoy_quic_stream.cc]: Whoops, thanks! (cut-and-paste error from an earlier version where the encode was a member variable not a local variable).

File: source/common/quic/envoy_quic_stream.cc
Comment (RyanTheOptimist) [reply source/common/quic/envoy_quic_stream.cc:160]: Hm. Mostly because we have the same ASSERT in encodeHeaders and friends. I assumed it was valid because if `local_send_stream_` were true, then so would `write_side_closed()` and we would early return from this method on line 158. WDYT?

File: source/common/quic/envoy_quic_stream.cc
Comment (RyanTheOptimist) [reply source/common/quic/envoy_quic_stream.cc:172]: Done

File: source/common/quic/envoy_quic_stream.cc
Comment (danzh2010) [inline source/common/quic/envoy_quic_stream.cc:183]: should be !connected()?

Comment (RyanTheOptimist) [reply source/common/quic/envoy_quic_stream.cc]: Doh! Fixed.

File: test/integration/multiplexed_integration_test.cc
Comment (danzh2010) [inline test/integration/multiplexed_integration_test.cc:1036]: The comment seems irrelevant.

File: source/common/quic/envoy_quic_stream.cc
Comment (danzh2010) [inline source/common/quic/envoy_quic_stream.cc:175]: Can this be true if the metadata passed in is empty?

Comment (RyanTheOptimist) [reply source/common/quic/envoy_quic_stream.cc:175]: No, because the METADATA frame header will always be present.

**Review by danzh2010 — APPROVED** [review]

File: source/common/quic/envoy_quic_client_stream.cc
Comment (alyssawilk) [inline source/common/quic/envoy_quic_client_stream.cc:415]: in H3, can metadata arrive before headers?  I think it's an invariant of Envoy that headers arrive first and I'm not sure if it's an invariant of the quiche libraries.

File: source/common/quic/envoy_quic_stream.cc
Comment (alyssawilk) [inline source/common/quic/envoy_quic_stream.cc:151]: update comment and log: not supported by config

File: source/common/quic/envoy_quic_stream.cc
Comment (alyssawilk) [inline source/common/quic/envoy_quic_stream.cc:180]: is bad application payload the best?  feels like an invariant failure but idk if ther's a better code.

File: source/common/quic/envoy_quic_server_stream.cc
Comment (alyssawilk) [inline source/common/quic/envoy_quic_server_stream.cc:455]: do we need to do any validation steps with or without UHV? I honestly don't recall sorry

File: test/integration/multiplexed_integration_test.cc
Comment (alyssawilk) [inline test/integration/multiplexed_integration_test.cc:1003]: can we change Http2MetadataIntegrationTest to be MetadataIntegrationTest and run all the tests for H2/H3?

**Review by alyssawilk — COMMENTED** [review]
Nice addition!
/wait on comments

File: test/integration/multiplexed_integration_test.cc
Comment (RyanTheOptimist) [reply test/integration/multiplexed_integration_test.cc]: Done!

**Comment by RyanTheOptimist** [comment]
@alyssawilk PTAL. The QUICHE update which the size limit tests require will be landed in #32892. In the meantime, I've patched that .bzl file change into this PR so I'll need a main merge once that lands. But it's ready for review, I think!

File: source/common/quic/envoy_quic_client_stream.cc
Comment (alyssawilk) [reply source/common/quic/envoy_quic_client_stream.cc:415]: I think we should either make it an error, or add an e2e test it doesn't cause problems in Envoy.

File: source/common/quic/envoy_quic_server_stream.cc
Comment (alyssawilk) [reply source/common/quic/envoy_quic_server_stream.cc]: Ok I didn't know if we had concerns about character validation, since I think Envoy has some assumptions there, but maybe not for metadata.
cc @yanavlasov

File: test/integration/multiplexed_integration_test.cc
Comment (alyssawilk) [inline test/integration/multiplexed_integration_test.cc:558]: comment at least the first of these?  I assume this is one of those "H3 sending zero byte chunk" differences?

File: test/integration/multiplexed_integration_test.cc
Comment (alyssawilk) [inline test/integration/multiplexed_integration_test.cc:591]: optional, switch to adding expect 100continue to default_request_headers_

File: test/integration/multiplexed_integration_test.cc
Comment (alyssawilk) [inline test/integration/multiplexed_integration_test.cc:911]: default request headers here and below?

File: test/integration/multiplexed_integration_test.cc
Comment (alyssawilk) [inline test/integration/multiplexed_integration_test.cc:287]: Http2MetadataIntegrationTest -> MetadataIntegraionTest?

**Review by alyssawilk — COMMENTED** [review]
looks like test/integration/protocol_integration_test.cc has some Metadata tests as well as http2_flood_integration_test.cc

Given it's hidden and we're the only users I'm OK with a TODO to turn those up and landing to kick off import - your call if that's helpful or if you want to just land the right proper fix :-)

File: test/integration/multiplexed_integration_test.cc
Comment (RyanTheOptimist) [reply test/integration/multiplexed_integration_test.cc:287]: Whoops, done!

File: test/integration/multiplexed_integration_test.cc
Comment (RyanTheOptimist) [reply test/integration/multiplexed_integration_test.cc:560]: Done. (Commented all of the places to be consistent.)

File: test/integration/multiplexed_integration_test.cc
Comment (RyanTheOptimist) [reply test/integration/multiplexed_integration_test.cc]: Done.

File: test/integration/multiplexed_integration_test.cc
Comment (RyanTheOptimist) [reply test/integration/multiplexed_integration_test.cc]: Done.

File: source/common/quic/envoy_quic_client_stream.cc
Comment (RyanTheOptimist) [reply source/common/quic/envoy_quic_client_stream.cc:415]: Oh! I think we have a test for this already! `ProxyMetadataInResponse` does:

```
  // Sends metadata before response header.                                                                                                                                                 
  const std::string key = "key";
  std::string value = std::string(80 * 1024, '1');
  Http::MetadataMap metadata_map = {{key, value}};
  Http::MetadataMapPtr metadata_map_ptr = std::make_unique<Http::MetadataMap>(metadata_map);
  Http::MetadataMapVector metadata_map_vector;
  metadata_map_vector.push_back(std::move(metadata_map_ptr));
  upstream_request_->encodeMetadata(metadata_map_vector);
  upstream_request_->encodeHeaders(default_response_headers_, false);
  upstream_request_->encodeData(12, true);
```

File: source/common/quic/envoy_quic_server_stream.cc
Comment (RyanTheOptimist) [reply source/common/quic/envoy_quic_server_stream.cc]: @birenroy might also know

**Comment by RyanTheOptimist** [comment]
> looks like test/integration/protocol_integration_test.cc has some Metadata tests as well as http2_flood_integration_test.cc
> 
> Given it's hidden and we're the only users I'm OK with a TODO to turn those up and landing to kick off import - your call if that's helpful or if you want to just land the right proper fix :-)

Thanks! Yeah, if it's ok with you, I'd love to address this in a follow up. Hopefully it's straightforward :)

**Review by RyanTheOptimist — COMMENTED** [review]
Thanks for the comments! The QUICHE merge landed so this PR no longer includes it.

**Review by alyssawilk — APPROVED** [review]
mildly prefer an explicit TODO for the rest of tests but realistically I know you're good for it =P

File: source/common/quic/envoy_quic_server_stream.cc
Comment (birenroy) [reply source/common/quic/envoy_quic_server_stream.cc]: No restrictions on metadata character set, since it is opt-in and the strings are length prefixed.

**Pull request merged**
